inc/msp/Client.hpp

```diff
@@ -16,210 +16,364 @@
+    bool sendMessage(msp::Message& message, const double& timeout = 0);
     
+    /**
+     * @brief Send a message, but do not wait for any response
+     * @param message Reference to a Message-derived object to be sent
+     */
+    bool sendMessageNoWait(const msp::Message& message);
     
-
-
     /**
-     * @brief subscribe register callback function that is called when type is received
-     * @param callback pointer to callback function (class method)
-     * @param context class with callback method
-     * @param tp period of timer that will send subscribed requests (in seconds), by default this is 0 and requests are not sent periodically
+     * @brief Register callback function that is called when a message of 
+     * matching ID is received
+     * @param callback Pointer to callback function (class method)
+     * @param context Object containing callback method
+     * @param tp Period of timer that will send subscribed requests (in seconds). 
      * @return pointer to subscription that is added to internal list
      */
     template<typename T, typename C, class = typename std::enable_if<std::is_base_of<msp::Message, T>::value>::type>
-    std::shared_ptr<SubscriptionBase> subscribe(void (C::*callback)(T&), C *context, const double tp = 0.0) {
+    std::shared_ptr<SubscriptionBase> subscribe(void (C::*callback)(const T&), C *context, const double& tp) {
         return subscribe<T>(std::bind(callback, context, std::placeholders::_1), tp);
     }
 
     /**
-     * @brief subscribe register callback function that is called when type is received
-     * @param callback function (e.g. lambda, class method, function pointer)
-     * @param tp period of timer that will send subscribed requests (in seconds), by default this is 0 and requests are not sent periodically
+     * @brief Register callback function that is called when a 
+     * message of matching ID is received
+     * @param recv_callback Function to be called upon receipt of message
+     * @param tp Period of timer that will send subscribed requests (in seconds). 
      * @return pointer to subscription that is added to internal list
      */
     template<typename T, class = typename std::enable_if<std::is_base_of<msp::Message, T>::value>::type>
-    std::shared_ptr<SubscriptionBase> subscribe(const std::function<void(T&)> &recv_callback, const double tp = 0.0) {
-
-        if(!(tp>=0.0))
-            throw std::runtime_error("Period must be positive!");
-
+    std::shared_ptr<SubscriptionBase> subscribe(const std::function<void(const T&)> &recv_callback, const double& tp) {
+        //validate the period
+        if(!(tp>=0.0)) throw std::runtime_error("Period must be positive!");
+        //get the id of the message in question
         const msp::ID id = T(fw_variant).id();
-        std::cout << "SUBSCRIBING TO " << (uint32_t)id << std::endl;
-        
-        
+        if (log_level_ >= INFO) std::cout << "SUBSCRIBING TO " << (uint32_t)id << std::endl;
```

Comment:
I guess you meant `(uint16_t)id` here? I.e. convert it to the raw 16bit MSP2 ID.
Btw, can you use casting like `uint16_t(id)` instead of the old-style cast? This is just for reducing compiler warnings.